src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs

```diff
@@ -42,23 +42,30 @@ public override async Task ComputeRefactoringsAsync(CodeRefactoringContext conte
             var cancellationToken = context.CancellationToken;
             var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
 
-            var identifier = await root.SyntaxTree.GetTouchingTokenAsync(context.Span.Start,
-                token => token.Parent.IsKind(SyntaxKind.LocalFunctionStatement), cancellationToken).ConfigureAwait(false);
-            if (identifier == default)
+            LocalFunctionStatementSyntax localFunction = default;
+            if (context.Span.IsEmpty)
             {
-                return;
+                // This branch is required to handle "C LocalFunction[||](C c)" -> root.FindNode return ParameterList
+                var identifier = await root.SyntaxTree.GetTouchingTokenAsync(context.Span.Start, token => token.Parent.IsKind(SyntaxKind.LocalFunctionStatement), cancellationToken).ConfigureAwait(false);
+                if (identifier == default)
+                {
+                    return;
+                }
+                localFunction = (LocalFunctionStatementSyntax)identifier.Parent;
             }
-
-            if (context.Span.Length > 0 &&
-                context.Span != identifier.Span)
+            else
             {
-                return;
-            }
+                var enclosingNode = root.FindNode(context.Span);
+                if (!(enclosingNode is LocalFunctionStatementSyntax))
+                {
+                    return;
+                }
+                localFunction = (LocalFunctionStatementSyntax)enclosingNode;
 
-            var localFunction = (LocalFunctionStatementSyntax)identifier.Parent;
-            if (localFunction.Identifier != identifier)
-            {
-                return;
+                if (context.Span != localFunction.Span && context.Span != localFunction.Identifier.Span)
+                {
+                    return;
+                }
```

Comment:
this check is likely too narrow.  it will only match the exact amount being selected.  however, users could totally do something like this:

```c#
[|        void Whatever()
        {
        }
|]
```

expecting that to work ok.

Reply:
basically we should write a helper here.  here's how i think it should work:

1. the span start should be between the end of the previous token's last non-whitespace trivia and the start of the node you care about.
2. the span should end between the end of your node and the start of the first non-whitespace trivia of the next token.

Reply:
Yep, I plan to create a helper for this kind of situation. Wanted to go through more problematic refactorings before deciding on how it should behave first. Will definitely fix, however.

Reply:
Note: we have a CodeRefactoringHelpers class.  For this specific helper, i would place it there.

Reply:
In regards to the helper, don't we also want to handle cases like these:
```csharp
      vo[|id Whatever()
        {
        }
|]
```

Reply:
I feel we don't need any checks on top of `FindNode` test - if that method returns a non-null local function syntax, user has partially or fully selected the local function. Partial selection covers one or more child tokens and nodes, so the user intention is not to select any specific child node or token and we should offer all local function specific refactorings.

Reply:
I believe following steps should suffice for this refactoring and can be generalized to all refactorings that logically work on a very specific syntax node:

1. Invoke FindNode and check if returned node is local function syntax. This should handle following 2 cases:
   1. Non-empty selection where user meant to select the local function and did it partially or fully.
   2. Empty selection where caret is inside a token whose immediate parent node is a local function.
2. Otherwise, if FindNode didn't succeed and selection is empty, caret might be between 2 tokens. Check if [FindTokenOnLeftOfPosition](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Shared/Extensions/SyntaxNodeExtensions.cs,647) or [FindTokenOnRightOfPosition](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Shared/Extensions/SyntaxNodeExtensions.cs,608) returns a token whose Parent node is a local function.

Reply:
Ok, that could actually be a very useful helper (`public static async Task<TNode> TryGetSelectedNode<TNode>(TextSpan selectedSpan, SyntaxNode root, CancellationToken cancellationToken) where TNode : SyntaxNode`). I'll write it and ?put it to CodeREfatoringsHelpers.

As for using `FindTokenOnLeftOfPosition`, is there any benefit over `GetTouchingTokenAsync`?

And lastly, we still need helper to eat whitespace around selection so that FindNode finds the node correctly even for larger selections (IIRC, will check to which token is whitespace actually attributed).

Reply:
> As for using FindTokenOnLeftOfPosition, is there any benefit over GetTouchingTokenAsync

It wasn't very clear to me which token it will return if caret is in between two tokens, and FindxxxLeft and FindxxxRight seemed to have much more clear semantics and doc comments. I am fine with you choosing GetTouchingTokenAsync or some other helper, as long as the logic can be kept simple.

> And lastly, we still need helper to eat whitespace around selection so that FindNode finds the node correctly even for larger selections (IIRC, will check to which token is whitespace actually attributed).

Can you test if FindNode already doesn't take care of this? My assumption is FindNode is designed to map spans to corresponding relevant node, so it is possible that it handles this case. If not, your point is correct about need for additional checks.

Reply:
FindNode doesn't do any processing on the given span and so we need a helper to truncate whitespace before & after.

Reply:
Sounds good, thanks!

Reply:
My bad, FindNode uses FullSpan that actually contains both whitespaces (pre, post).

Reply:
Ok, taking that back. FullSpan contains some before/after whitespace but doesn't really like newspaces.

Reply:
Stripping was required in the end in cases where there're newlines after the local function selected.